### PR TITLE
VBM-96: Phone verification

### PR DESCRIPTION
**Pull request by VladimirBludov**

### CODE REVIEW

File: packages/mobile/src/components/header-button/header-button.tsx [hunk 1]
```diff
@@ -19,13 +20,16 @@ const HeaderButton: FC<Props> = ({ label, onPress }) => {
         globalStyles.justifyContentStart,
       ]}
     >
-      <BackIcon size={35} style={{ color: ColorPalette.WHITE_100 }} />
+      <BackIcon
+        size={35}
+        style={{ color: color ? color : ColorPalette.WHITE_100 }}
```
Comment (seal323) [inline packages/mobile/src/components/header-button/header-button.tsx:25]:
```suggestion
        style={{ color: color ||ColorPalette.WHITE_100 }}
```

File: packages/mobile/src/components/header-button/header-button.tsx [hunk 2]
```diff
@@ -19,13 +20,16 @@ const HeaderButton: FC<Props> = ({ label, onPress }) => {
...
+      />
       <Text
         style={[
           styles.buttonText,
           globalStyles.fs17,
           globalStyles.fontWeightMedium,
-          { color: ColorPalette.WHITE_100, textAlign: 'left' },
+          { color: color ? color : ColorPalette.WHITE_100, textAlign: 'left' },
```
Comment (seal323) [inline packages/mobile/src/components/header-button/header-button.tsx:32]:
```suggestion
          { color: color || ColorPalette.WHITE_100, textAlign: 'left' },
```

File: packages/mobile/src/components/header-button/header-button.tsx (hunk 1 above)
Comment (VladimirBludov) [reply packages/mobile/src/components/header-button/header-button.tsx]: corrected

File: packages/mobile/src/components/header-button/header-button.tsx (hunk 2 above)
Comment (VladimirBludov) [reply packages/mobile/src/components/header-button/header-button.tsx]: corrected

File: packages/mobile/src/screens/verify-phone-screens/components/container/container.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React, { FC, ReactNode } from 'react';
+import { View } from '~/components/components';
+import { globalStyles } from '~/styles/styles';
+
+type Props = {
+  children: ReactNode;
+};
+
+const Container: FC<Props> = ({ children }) => {
+  return <View style={globalStyles.px5}>{children}</View>;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/components/container/container.tsx:10]: Looks redundant to me.

File: packages/mobile/src/screens/verify-phone-screens/components/header/header.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+type Props = {
+  labelButton?: string;
+  onPress?: () => void;
+  hideButton?: boolean;
+};
+
+const Header: FC<Props> = ({ labelButton, onPress, hideButton }) => {
+  const { t } = useTranslation();
+  const { colors } = useCustomTheme();
+  const isShowButton = !hideButton && onPress && labelButton;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/components/header/header.tsx:16]:
```suggestion
  const isButtonVisible = !hideButton && onPress && labelButton;
```

File: packages/mobile/src/screens/verify-phone-screens/components/header/styles.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const { width } = Dimensions.get('window');
+
+const styles = StyleSheet.create({
+  header: {
+    height: 64,
+    paddingBottom: 10,
+  },
+  text: {
+    position: 'absolute',
+    zIndex: -1,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/components/header/styles.ts:12]: Can we avoid absolute positioning here?

File: packages/mobile/src/screens/verify-phone-screens/components/title/title.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import React, { FC } from 'react';
+import { TextStyle, ViewStyle } from 'react-native';
+import { useCustomTheme } from '~/hooks/hooks';
+import { Text, View } from '~/components/components';
+import { globalStyles } from '~/styles/styles';
+import { styles } from './styles';
+
+type Props = {
+  label: string;
+  contentStyle?: TextStyle;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/components/title/title.tsx:10]:
```suggestion
  textStyle?: StyleProp<TextStyle>;
```

File: packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  const handleContinuePress = (): void => {
+    navigation.navigate(RootScreenName.MAIN, {
+      screen: MainScreenName.ACCOUNT_ROOT,
+    });
+  };
+
+  return (
+    <Wrapper>
+      <Header hideButton />
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx:35]:
```suggestion
      <Header hideButton={true} />
```

File: packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      <KeyboardAvoiding>
+        <Container>
+          <VerifyImage
+            source={images.verified}
+            contentContainerStyle={globalStyles.mt6}
+          />
+          <Title
+            label={t('verificationPhone.VERIFIED_TITLE')}
+            contentContainerStyle={{ ...globalStyles.mt6, marginTop: 75 }}
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx:44]:
```suggestion
            contentContainerStyle={[globalStyles.mt6, { marginTop: 75 }]}
```

But generally you should avoid inline styles

File: packages/mobile/src/screens/verify-phone-screens/verify-code/styles.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { StyleSheet, Dimensions } from 'react-native';
+
+const { width } = Dimensions.get('window');
+
+const styles = StyleSheet.create({
+  buttonContainer: {
+    width: width / 2.3,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/styles.ts:7]: Can we avoid using magic numbers?

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  const userPhone = useAppSelector(selectUserPhone);
+  const dataStatus = useAppSelector(selectUserActionDataStatus);
+  const isLoading = dataStatus === DataStatus.PENDING;
+  const { control, errors, handleSubmit } = useAppForm<PhoneVerifyDto>({
+    defaultValues: {
+      code: '',
+    },
+  });
+
+  const handleBackButton = (): void => {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:54]:
```suggestion
  const handleBackButtonPress = (): void => {
```

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      });
+  };
+
+  const onSubmit = (payload: PhoneVerifyDto): void => {
+    dispatch(authActions.verifyPhone(payload))
+      .unwrap()
+      .then(() => {
+        navigation.navigate(RootScreenName.VERIFY, {
+          screen: VerifyScreenName.VERIFIED,
+        });
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:72]:
```suggestion
        }).catch();
```

File: packages/mobile/src/services/auth-api/auth-api.service.ts [hunk 12]
```diff
@@ -47,6 +50,51 @@ class AuthApi {
...
+        contentType: HttpContentType.APPLICATION_JSON,
+        params: {
+          userId: _payload,
+        },
+        hasAuth: true,
+      },
+    );
+  }
+
+  verifyPhone(_payload: PhoneVerifyDto, _userId: string): Promise<unknown> {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/auth-api/auth-api.service.ts:84]: These belongs to a separate service

File: packages/mobile/src/store/auth/common.ts [hunk 13]
```diff
@@ -1,7 +1,10 @@
 enum ActionType {
   SIGN_IN = 'auth/sign-in',
   SIGN_UP = 'auth/sign-up',
+  CURRENT_USER = 'auth/current-user',
   LOG_OUT = 'auth/log-out',
+  SEND_CODE = 'verify/send-code',
+  VERIFY_PHONE = 'verify/verify-phone',
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/auth/common.ts:7]: Create additional actions for that

File: packages/mobile/src/common/types/navigation/navigation-props.ts [hunk 14]
```diff
@@ -1,7 +1,11 @@
 import { NativeStackNavigationProp } from '@react-navigation/native-stack';
-
-import { RootNavigationParamList } from '~/common/types/navigation/navigation';
+import {
+  RootNavigationParamList,
+  VerifyNavigationParamList,
+} from '~/common/types/navigation/navigation';
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/common/types/navigation/navigation-props.ts:5]: corrected

File: packages/mobile/src/common/types/navigation/root-navigation-param-list.type.ts [hunk 15]
```diff
@@ -1,12 +1,13 @@
 import { RootScreenName } from '~/common/enums/enums';
 
 type RootNavigationParamList = {
-  [RootScreenName.MAIN]: undefined;
+  [RootScreenName.MAIN]?: { screen: string };
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/common/types/navigation/root-navigation-param-list.type.ts:4]: corrected

File: packages/mobile/src/common/types/verify/verify.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+type VerifyPhone = {
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/common/types/verify/verify.ts:1]: corrected

File: packages/mobile/src/localization/languages/en/screens.ts [hunk 17]
```diff
@@ -81,12 +81,30 @@ export const screens = {
...
+    INPUT_LABEL_PHONE: 'Phone',
+    VERIFY_LATER: 'Verify later',
+    BACK_HOME: 'Home',
+    BACK_BUTTON: 'Back',
+    ENTER_CODE: 'Enter a code',
+    JUST_SENT: 'We just sent a code to',
+    INPUT_LABEL_CODE: 'Enter the verification code we just sent to you',
+    RESEND_CODE: 'Resend code',
+    CONTINUE: 'Continue',
+    VERIFIED_TITLE: 'Your account has fully been verified',
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/localization/languages/en/screens.ts:99]: corrected

File: packages/mobile/src/navigation/navigation.tsx [hunk 18]
```diff
@@ -17,6 +17,7 @@ import {
 } from '~/screens/screens';
 import { HeaderLeft, HeaderSave } from '~/components/components';
 import { MainNavigation } from './tabs/tabs.navigation';
+import { VerifyNavigation } from './verify/verify.navigation';
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/navigation/navigation.tsx:20]: corrected

File: packages/mobile/src/navigation/verify/verify.navigation.tsx [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+} from '@react-navigation/native-stack';
+import { VerifyScreenName } from '~/common/enums/enums';
+import { VerifyNavigationParamList } from '~/common/types/types';
+import {
+  VerifyPhoneScreen,
+  VerifyCodeScreen,
+  VerifiedScreen,
+} from '~/screens/screens';
+
+const VerifyNavigation: FC = () => {
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/navigation/verify/verify.navigation.tsx:14]: corrected

File: packages/mobile/src/screens/auth/auth.tsx [hunk 20]
```diff
@@ -41,7 +42,11 @@ const Auth: FC = () => {
   };
 
   const handleSignUp = (payload: UserSignUpDto): void => {
-    dispatch(authActions.signUp(payload));
+    dispatch(authActions.signUp(payload))
+      .unwrap()
+      .then(() => {
+        navigation.navigate(RootScreenName.VERIFY);
+      });
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/screens/auth/auth.tsx:49]: corrected

File: packages/mobile/src/screens/verify-phone-screens/components/container/container.tsx (hunk 3 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/components/container/container.tsx:10]: corrected

File: packages/mobile/src/screens/verify-phone-screens/components/header/header.tsx (hunk 4 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/components/header/header.tsx:16]: corrected

File: packages/mobile/src/screens/verify-phone-screens/components/header/styles.ts (hunk 5 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/components/header/styles.ts:12]: corrected

File: packages/mobile/src/screens/verify-phone-screens/components/title/title.tsx (hunk 6 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/components/title/title.tsx:10]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx (hunk 7 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx:35]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx (hunk 8 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verified/verified.tsx:44]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/styles.ts (hunk 9 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/styles.ts:7]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx (hunk 10 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:54]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx (hunk 11 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:72]: corrected

File: packages/mobile/src/services/auth-api/auth-api.service.ts (hunk 12 above)
Comment (VladimirBludov) [reply packages/mobile/src/services/auth-api/auth-api.service.ts:84]: corrected

File: packages/mobile/src/store/auth/common.ts (hunk 13 above)
Comment (VladimirBludov) [reply packages/mobile/src/store/auth/common.ts:7]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+  const dispatch = useAppDispatch();
+  const navigation = useNavigation<RootNavigationProps>();
+  const { colors } = useCustomTheme();
+  const userPhone = useAppSelector(selectUserPhone);
+  const dataStatusVerify = useAppSelector(selectVerifyPhoneDataStatus);
+  const dataStatusAuth = useAppSelector(selectAuthDataStatus);
+  const [isValidCode, setIsValidCode] = useState(false);
+  const isLoading =
+    dataStatusVerify === DataStatus.PENDING ||
+    dataStatusAuth === DataStatus.PENDING;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:56]:
```suggestion
  const isLoading = [dataStatusVerify, dataStatusAuth].includes(DataStatus.PENDING);
```

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+  };
+
+  const handleResendPress = (): void => {
+    dispatch(verifyPhoneActions.sendCodeVerifyPhone())
+      .unwrap()
+      .then(() => {
+        notification.success(t('verificationPhone.CODE_SENT'));
+      })
+      .catch((err) => {
+        throw err;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:74]: Same here, unless you introduce some sort of error boundary
```suggestion
        console.warn(err);
```

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+  };
+
+  const handleChange = (text: string): unknown => {
+    if (text.match(CODE_REGEX)) {
+      setIsValidCode(true);
+
+      return;
+    }
+    setIsValidCode(false);
+  };
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:96]: Let's stick with one entry point for all form handlers

File: packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+  PhoneVerifyDto,
+} from '@vse-bude/shared';
+import { Http } from '../http/http.service';
+
+type Constructor = {
+  http: Http;
+  apiPrefix: string;
+};
+
+class VerifyPhoneApi {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts:15]:
```suggestion
class PhoneVerificationApi {
```

File: packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+  #http: Http;
+
+  #apiPrefix: string;
+
+  constructor({ http, apiPrefix }: Constructor) {
+    this.#http = http;
+    this.#apiPrefix = apiPrefix;
+  }
+
+  sendCodeVerifyPhone(): Promise<boolean> {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts:25]:
```suggestion
  sendVerificationCode(): Promise<boolean> {
```

Don't we need to specify payload (code itself) as well?

File: packages/mobile/src/validation-schemas/phone/phone.ts [hunk 26]
```diff
@@ -0,0 +1,17 @@
+import * as Joi from 'joi';
+import i18next from 'i18next';
+import { PHONE_NUMBER_REGEX } from '~/common/regexp/regexp';
+import { VerifyPhoneRequestDto } from '~/common/types/types';
+
+const phone = Joi.object<VerifyPhoneRequestDto>({
+  phone: Joi.string()
+    .pattern(PHONE_NUMBER_REGEX)
+    .trim()
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/validation-schemas/phone/phone.ts:9]:
```suggestion
    .trim()
    .pattern(PHONE_NUMBER_REGEX)
```

File: packages/mobile/src/services/services.ts [hunk 27]
```diff
@@ -31,8 +32,21 @@ const productApi = new ProductService({
   apiPrefix: ENV.APP.API_ORIGIN_URL,
 });
 
+const verifyPhoneApi = new VerifyPhoneApi({
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/services.ts:35]:
```suggestion
const phoneVerificationApi = new VerifyPhoneApi({
```

File: packages/mobile/src/common/regexp/validation/code.ts [hunk 28]
```diff
@@ -0,0 +1 @@
+export const CODE_REGEX = /^[0-9]{6}$/;
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/common/regexp/validation/code.ts:1]: corrected

File: packages/mobile/src/screens/auth/auth.tsx [hunk 29]
```diff
@@ -41,7 +42,14 @@ const Auth: FC = () => {
...
 
   const handleSignUp = (payload: UserSignUpDto): void => {
-    dispatch(authActions.signUp(payload));
+    dispatch(authActions.signUp(payload))
+      .unwrap()
+      .then(() => {
+        navigation.navigate(RootScreenName.VERIFY_PHONE);
+      })
+      .catch((err) => {
+        throw err;
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/screens/auth/auth.tsx:51]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx (hunk 21 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:56]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx (hunk 22 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:74]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx (hunk 23 above)
Comment (VladimirBludov) [reply packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:96]: corrected

File: packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts (hunk 24 above)
Comment (VladimirBludov) [reply packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts:15]: corrected

File: packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts (hunk 25 above)
Comment (VladimirBludov) [reply packages/mobile/src/services/verify-phone-api/verify-phone-api.service.ts:25]: no it's for getting the code

File: packages/mobile/src/validation-schemas/phone/phone.ts (hunk 26 above)
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/phone/phone.ts:9]: corrected

File: packages/mobile/src/services/services.ts (hunk 27 above)
Comment (VladimirBludov) [reply packages/mobile/src/services/services.ts:35]: corrected

File: packages/mobile/src/screens/personal-info/components/personal-info-form/personal-info-form.tsx [hunk 30]
```diff
@@ -31,7 +37,9 @@ type Props = {
 
 const PersonalInfoForm: React.FC<Props> = ({ onSubmit }) => {
   const { t } = useTranslation();
+  const navigation = useNavigation<RootNavigationProps>();
   const user = useAppSelector(selectCurrentUser);
+  const verified = useAppSelector(selectPhoneVerified);
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/screens/personal-info/components/personal-info-form/personal-info-form.tsx:42]: corrected

File: packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+              errors={errors}
+              contentContainerStyle={globalStyles.mt6}
+            />
+            {/* {!errors?.code && (
+              <CheckCircleIcon
+                style={styles.icon}
+                size={20}
+                color={colors.accent}
+              />
+            )} */}
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/verify-phone-screens/verify-code/verify-code.tsx:124]: No unused code should be left

**Review by AleksandrTermenzhy — APPROVED** [review]

**Review by seal323 — APPROVED** [review]
Merge after comment is fixed

**Pull request merged**